### PR TITLE
feat: support update in couchdb

### PR DESCRIPTION
**Pull request by fqutishat**
Signed-off-by: Firas Qutishat <firas.qutishat@securekey.com>

### CODE REVIEW

File: pkg/storage/couchdb/couchdbstore.go [hunk 1]
```diff
@@ -130,6 +130,8 @@ func (p *Provider) Close() error {
 // CouchDBStore represents a CouchDB-backed database.
 type CouchDBStore struct {
 	db *kivik.DB
+	// TODO change to leveldb
```
Comment (DRK3) [inline pkg/storage/couchdb/couchdbstore.go]: I'd suggest making an issue for it and linking it here in the comment

File: pkg/storage/couchdb/couchdbstore_test.go [hunk 2]
```diff
@@ -208,6 +208,9 @@ func TestCouchDBStore_Get(t *testing.T) {
 		value, err := store.Get(testDocKey)
 		require.NoError(t, err)
 		require.Equal(t, testJSONValue, string(value))
+
+		err = store.Put(testDocKey, []byte(testJSONValue))
```
Comment (DRK3) [inline pkg/storage/couchdb/couchdbstore_test.go]: Is this testing the update functionality? Don't you want to check and see if the _rev number is updated correctly?

Comment (fqutishat) [reply pkg/storage/couchdb/couchdbstore_test.go]: yes

File: pkg/storage/couchdb/couchdbstore.go [hunk 3]
```diff
@@ -141,11 +143,28 @@ func (c *CouchDBStore) Put(k string, v []byte) error {
...
+		}
+
+		m["_rev"] = c.revIDs[k]
+
+		var err error
+		valueToPut, err = json.Marshal(m)
+		if err != nil {
+			return fmt.Errorf("failed to marshal put value: %w", err)
+		}
+	}
```
Comment (DRK3) [inline pkg/storage/couchdb/couchdbstore.go]: Does this work with nested JSON?

File: pkg/storage/couchdb/couchdbstore.go [hunk 4]
```diff
@@ -141,11 +143,28 @@ func (c *CouchDBStore) Put(k string, v []byte) error {
...
 	}
 
-	_, err := c.db.Put(context.Background(), k, valueToPut)
+	if c.revIDs[k] != "" {
+		var m map[string]interface{}
+		if err := json.Unmarshal(valueToPut, &m); err != nil {
+			return fmt.Errorf("failed to unmarshal put value: %w", err)
+		}
+
+		m["_rev"] = c.revIDs[k]
```
Comment (DRK3) [inline pkg/storage/couchdb/couchdbstore.go]: Shouldn't this increment the revision number? Or change it somehow?

File: pkg/storage/couchdb/couchdbstore.go (hunk 3 above)
Comment (fqutishat) [reply pkg/storage/couchdb/couchdbstore.go]: yes

File: pkg/storage/couchdb/couchdbstore.go (hunk 4 above)
Comment (fqutishat) [reply pkg/storage/couchdb/couchdbstore.go]: no

File: pkg/storage/couchdb/couchdbstore_test.go (hunk 2 above)
Comment (fqutishat) [reply pkg/storage/couchdb/couchdbstore_test.go]: done

File: pkg/storage/couchdb/couchdbstore.go (hunk 1 above)
Comment (fqutishat) [reply pkg/storage/couchdb/couchdbstore.go]: done

File: pkg/storage/couchdb/couchdbstore.go [hunk 5]
```diff
@@ -141,11 +143,29 @@ func (c *CouchDBStore) Put(k string, v []byte) error {
...
+			return fmt.Errorf("failed to marshal put value: %w", err)
+		}
+	}
+
+	rev, err := c.db.Put(context.Background(), k, valueToPut)
 	if err != nil {
 		return err
 	}
 
+	c.revIDs[k] = rev
```
Comment (DRK3) [inline pkg/storage/couchdb/couchdbstore.go]: Oh, just saw this.

**Pull request merged**